File: app/src/main/java/com/example/newsapp/data/DataRepository/ArticleRepository.java
package com.example.newsapp.data.DataRepository;

import android.app.Application;
import android.arch.lifecycle.LiveData;
import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.AsyncTask;
import android.os.Handler;
import android.util.Log;
import android.widget.Toast;

import java.util.HashMap;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import com.example.newsapp.data.api.ApiManager;
import com.example.newsapp.data.database.AppDatabase;
import com.example.newsapp.data.database.accessobjects.ArticleAccessObject;
import com.example.newsapp.data.database.accessobjects.SourcesAccessObject;
import com.example.newsapp.data.datamodels.Articles;
import com.example.newsapp.data.datamodels.CountryConstants;
import com.example.newsapp.data.datamodels.DataResponse;
import com.example.newsapp.data.datamodels.Sources;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class ArticleRepository {

    private ArticleAccessObject articleAccessObject;
    private SourcesAccessObject sourcesAccessObject;
    private LiveData<List<Articles>> mAllArticles;
    private LiveData<List<Articles>> mAllArticlesByTitle;
    private LiveData<List<Articles>> mAllArticlesByDomain;
    private LiveData<List<Articles>> mAllArticlesNoQuery;
    private LiveData<List<Sources>> mSources;
    private LiveData<List<Articles>> mArticlesByCountry;
    private LiveData<List<String>> mCountries;
    private LiveData<List<String>> mNewsCategories;
    private LiveData<List<Articles>> mNewsBySetToRead;
    private LiveData<List<Articles>> mNewsByFavorite;
    private LiveData<List<Articles>> mNewsByDomain;

    private LiveData<List<Sources>> mSourcesByNewsCategory;

    private List<Articles> mTopHeadlines,
            mTopHeadlinesByCountryCategory,
            mTopHeadlinesByCountry,
            mTopHeadlinesBySearch;

    private Application application;


    public ArticleRepository(Application application){
        initializeDatabaseValues(application);
        this.application = application;

    }

    public void initializeDatabaseValues(Application application){
        AppDatabase appDatabase = AppDatabase.getDatabase(application);
        articleAccessObject  = appDatabase.articleAccessObject();
        sourcesAccessObject = appDatabase.sourcesAccessObject();
        mSources = sourcesAccessObject.fetchAllData();
        mCountries = sourcesAccessObject.fetchCountryLists();
        mAllArticlesNoQuery = articleAccessObject.fetchAllData();
        mNewsCategories = sourcesAccessObject.fetchCategoryList();

    }

    public void initializeDatabaseValues(Application application, String query){

        AppDatabase database = AppDatabase.getDatabase(application);
        articleAccessObject  = database.articleAccessObject();
        sourcesAccessObject = database.sourcesAccessObject();
    }

    public LiveData<List<Articles>> getmAllArticles(){

        return mAllArticles;
    }
    public LiveData<List<Articles>> getAllArticlesNoQuery(){
        return mAllArticlesNoQuery;
    }

    public LiveData<List<Sources>> getSources(){

        return mSources;
    }

    public LiveData<List<Sources>> getSourcesByNewsCategory(String categoryQuery){

        mSourcesByNewsCategory = sourcesAccessObject.fetchNewsNamesByCategory(categoryQuery);
        return mSourcesByNewsCategory;
    }

    public LiveData<List<Articles>> getArticlesByTitle(Context context, String title){

        try {
            new insertDataFromTitle(context,title,articleAccessObject).execute();

        } finally {

            mAllArticlesByTitle = articleAccessObject.fetchAllData("%"+title+"%");
        }

        return mAllArticlesByTitle;

    }

    public LiveData<List<Articles>> getArticlesByDomain(Context context, String domain) {

        try {
            new insertDataFromDomain(context,domain,articleAccessObject).execute();

        } finally {

            Pattern pattern = Pattern.compile("(https?://)([^:^/]*)(:\\d*)?(.*)?");
            Matcher matcher = pattern.matcher(domain);
            matcher.find();
            domain  = matcher.group(2).substring(4);

            mAllArticlesByDomain = articleAccessObject
                    .fetchDataByDomain("%"+domain+"%");
        }

        return mAllArticlesByDomain;

    }

    public LiveData<List<Articles>> getDataByFavorite(){
        mNewsByFavorite = articleAccessObject.fetchDataByFavorite();
        return mNewsByFavorite;
    }
    public LiveData<List<Articles>> getDataBySetToRead(){
        mNewsBySetToRead = articleAccessObject.fetchDataByToRead();
        return mNewsBySetToRead;
    }


    public LiveData<List<String>> getCountries(){
        return mCountries;
    }

    public LiveData<List<String>> getNewsCategories() {
        return mNewsCategories;
    }

    public void insert(List<Articles> articles){
        new insertAsyncTask(articleAccessObject).execute(articles);

    }

    public void insertFavorite(boolean setToFavorite, String title){

        HashMap hashMap = new HashMap();
        hashMap.put("favorite",setToFavorite);
        hashMap.put("title", title);

        new insertFavoriteArticle(articleAccessObject).execute(hashMap);

    }
    public void insertSetToRead(boolean setToReadValue, String title){
        HashMap hashMap = new HashMap();
        hashMap.put("toReadValue",setToReadValue);
        hashMap.put("title", title);


        new insertSetToRead(articleAccessObject).execute(hashMap);

    }

    public LiveData<List<Articles>> getArticlesByCountry(Context context, String country) {

        try {

            CountryConstants countryConstants = new CountryConstants();

            HashMap countryMap = countryConstants.countryListData();

            for (Object key : countryMap.keySet()) {
                if (country.equals(countryMap.get(key))) {

                    new insertDataFromCountry(context,key.toString().toLowerCase()
                            ,articleAccessObject).execute();
                }
            }

        } finally {

            mArticlesByCountry = articleAccessObject
                    .fetchAllCountryData("%"+country+"%");
        }

        return mArticlesByCountry;
    }

    public static class insertAsyncTask extends AsyncTask<List<Articles>, Void, Void> {

        private ArticleAccessObject mAsyckTaskAccessObject;

        insertAsyncTask(ArticleAccessObject articleAccessObject){
            mAsyckTaskAccessObject = articleAccessObject;
        }

        @Override
        protected  Void doInBackground(final List<Articles>... params){

            for (int index = 0; index < params[0].size(); index++) {
                mAsyckTaskAccessObject.createDataIfNotExists(params[0].get(index));
            }

            return null;
        }

    }


    public static class insertDataFromCountry extends AsyncTask<String, Void, Void> {

        private ArticleAccessObject mInsertByCountryAccessObject;
        private Context context;
        private String queryValue;
        private List<Articles> articlesByCountry, modifiedArticleList;

        insertDataFromCountry(Context context, String query,ArticleAccessObject articleAccessObject){
            this.context = context;
            this.queryValue = query;
            mInsertByCountryAccessObject = articleAccessObject;
        }

        @Override
        protected Void doInBackground(String... strings){
            if (strings != null){

                ApiManager apiManager = new ApiManager(context, queryValue);

                ConnectivityManager connectivityManager =(ConnectivityManager)context
                        .getSystemService(Context.CONNECTIVITY_SERVICE);

                NetworkInfo activeNetwork = connectivityManager.getActiveNetworkInfo();

                boolean isConnected = activeNetwork != null &&
                        activeNetwork.isConnectedOrConnecting();

                if (isConnected) {

                    Call<DataResponse> call = apiManager.getTopHeadlines();
                    call.enqueue(new Callback<DataResponse>() {
                        @Override
                        public void onResponse(Call<DataResponse> call,
                                               Response<DataResponse> response) {

                            if (response.body() != null) {
                                CountryConstants countryConstants = new CountryConstants();
                                String countryQuery = countryConstants
                                        .countryListData().get(queryValue.toUpperCase());

                                articlesByCountry = response.body().getArticles();
                                for (int index = 0; index < articlesByCountry.size(); index++) {

                                    String author = articlesByCountry.get(index).getAuthor();
                                    articlesByCountry.get(index)
                                            .setAuthor(author + "\n"+ countryQuery);

                                    mInsertByCountryAccessObject
                                            .createDataIfNotExists(articlesByCountry
                                            .get(index));

                                }
                            }
                        }

                        @Override
                        public void onFailure(Call<DataResponse> call, Throwable t) {
                            Log.d("Data", t.getLocalizedMessage());

                        }
                    });
                } else {

                    Handler handler =  new Handler(context.getMainLooper());
                    handler.post( new Runnable(){
                        public void run(){
                            Toast.makeText(context,"You are not connected to the internet",
                                    Toast.LENGTH_LONG).show();
                        }
                    });
                }
            }
            return null;
        }

    }


    public static class insertDataFromTitle extends AsyncTask<String, Void, Void> {

        private ArticleAccessObject mInsertByTitleAccessObject;
        private Context context;
        private String queryValue;
        private List<Articles> articlesByTitle;

        insertDataFromTitle(Context context, String query,
                            ArticleAccessObject articleAccessObject){
            this.context = context;
            this.queryValue = query;
            mInsertByTitleAccessObject = articleAccessObject;
        }

        @Override
        protected Void doInBackground(String... strings){
            if (strings != null){

                ApiManager apiManager = new ApiManager(context, queryValue);

                ConnectivityManager connectivityManager =(ConnectivityManager)context
                        .getSystemService(Context.CONNECTIVITY_SERVICE);

                NetworkInfo activeNetwork = connectivityManager.getActiveNetworkInfo();

                boolean isConnected = activeNetwork != null &&
                        activeNetwork.isConnectedOrConnecting();

                if (isConnected) {

                    Call<DataResponse> call = apiManager.getArticles();
                    call.enqueue(new Callback<DataResponse>() {
                        @Override
                        public void onResponse(Call<DataResponse> call,
                                               Response<DataResponse> response) {

                            if (response.body() != null) {

                                articlesByTitle = response.body().getArticles();

                                for (int index = 0; index < articlesByTitle.size(); index++) {

                                    mInsertByTitleAccessObject.createDataIfNotExists(articlesByTitle
                                            .get(index));

                                }
                            }
                        }

                        @Override
                        public void onFailure(Call<DataResponse> call, Throwable t) {
                            Log.d("Data", t.getLocalizedMessage());

                        }
                    });
                } else {

                    Handler handler =  new Handler(context.getMainLooper());
                    handler.post( new Runnable(){
                        public void run(){
                            Toast.makeText(context,"You are not connected to the internet",
                                    Toast.LENGTH_LONG).show();
                        }
                    });
                }
            }
            return null;
        }

    }
    public static class insertDataFromDomain extends AsyncTask<String, Void, Void> {

        private ArticleAccessObject mInsertByTitleAccessObject;
        private Context context;
        private String queryValue;
        private List<Articles> articlesByDomain;

        insertDataFromDomain(Context context, String domain,ArticleAccessObject articleAccessObject){
            this.context = context;
            this.queryValue = domain;

            Pattern pattern = Pattern.compile("(https?://)([^:^/]*)(:\\d*)?(.*)?");
            Matcher matcher = pattern.matcher(domain);
            matcher.find();
            this.queryValue   = matcher.group(2).substring(4);
            mInsertByTitleAccessObject = articleAccessObject;
        }

        @Override
        protected Void doInBackground(String... strings){
            if (strings != null){

                ApiManager apiManager = new ApiManager(context, queryValue);

                ConnectivityManager connectivityManager =(ConnectivityManager)context
                        .getSystemService(Context.CONNECTIVITY_SERVICE);

                NetworkInfo activeNetwork = connectivityManager.getActiveNetworkInfo();

                boolean isConnected = activeNetwork != null &&
                        activeNetwork.isConnectedOrConnecting();

                if (isConnected) {

                    Call<DataResponse> call = apiManager.getNewsByDomains();
                    call.enqueue(new Callback<DataResponse>() {
                        @Override
                        public void onResponse(Call<DataResponse> call,
                                               Response<DataResponse> response) {

                            if (response.body() != null) {

                                articlesByDomain = response.body().getArticles();

                                for (int index = 0; index < articlesByDomain.size(); index++) {

                                    mInsertByTitleAccessObject.createDataIfNotExists(articlesByDomain
                                            .get(index));

                                }
                            }
                        }

                        @Override
                        public void onFailure(Call<DataResponse> call, Throwable t) {
                            Log.d("Data", t.getLocalizedMessage());

                        }
                    });
                } else {

                    Handler handler =  new Handler(context.getMainLooper());
                    handler.post( new Runnable(){
                        public void run(){
                            Toast.makeText(context,"You are not connected to the internet",
                                    Toast.LENGTH_LONG).show();
                        }
                    });
                }
            }
            return null;
        }

    }

    public static class insertFavoriteArticle extends AsyncTask<HashMap, Void, Void>{
        private ArticleAccessObject mFavoriteAsyncTaskAccessObject;

        insertFavoriteArticle(ArticleAccessObject articleAccessObject){
            mFavoriteAsyncTaskAccessObject = articleAccessObject;
        }
        @Override
        protected  Void doInBackground(HashMap... hashMaps){
            if (hashMaps != null) {

                boolean favorite = (boolean)hashMaps[0].get("favorite");
                String title = (String)hashMaps[0].get("title");
                mFavoriteAsyncTaskAccessObject.setNewsItemToFavorite(favorite,
                        title);
            }
            return null;
        }

    }
    public static class insertSetToRead extends AsyncTask<HashMap, Void, Void>{
        private ArticleAccessObject mSetToReadAsyncTaskAccessObject;

        insertSetToRead(ArticleAccessObject articleAccessObject){
            mSetToReadAsyncTaskAccessObject = articleAccessObject;
        }
        @Override
        protected  Void doInBackground(HashMap ... hashMaps){
            if (hashMaps != null) {

                boolean toReadValue = (boolean)hashMaps[0].get("toReadValue");
                String title = (String)hashMaps[0].get("title");
                mSetToReadAsyncTaskAccessObject.setNewsItemToRead(toReadValue, title);
            }
            return null;
        }
    }

}

File: app/src/main/java/com/example/newsapp/data/DataRepository/ArticleRepository.kt
package com.example.newsapp.data.DataRepository

import android.app.Application
import android.arch.lifecycle.LiveData
import android.content.Context
import android.net.ConnectivityManager
import android.os.AsyncTask
import com.example.newsapp.data.api.NetworkService
import com.example.newsapp.data.database.AppDatabase
import com.example.newsapp.data.database.accessobjects.ArticleAccessObject
import com.example.newsapp.data.database.accessobjects.SourcesAccessObject
import com.example.newsapp.data.datamodels.Articles
import com.example.newsapp.data.datamodels.Sources
import java.util.*
import java.util.regex.Pattern
import javax.inject.Inject

class ArticleRepository(private val application: Application) {

    private var articleAccessObject: ArticleAccessObject? = null
    private var sourcesAccessObject: SourcesAccessObject? = null
    private val mAllArticles: LiveData<List<Articles>>? = null
    private var mAllArticlesByTitle: LiveData<List<Articles>>? = null
    private var mAllArticlesByDomain: LiveData<List<Articles>>? = null
    var allArticlesNoQuery: LiveData<List<Articles>>? = null
        private set
    var sources: LiveData<List<Sources>>? = null
        private set
    private var mArticlesByCountry: LiveData<List<Articles>>? = null
    var countries: LiveData<List<String>>? = null
        private set
    var newsCategories: LiveData<List<String>>? = null
        private set
    private var mNewsBySetToRead: LiveData<List<Articles>>? = null
    private var mNewsByFavorite: LiveData<List<Articles>>? = null
    private val mNewsByDomain: LiveData<List<Articles>>? = null

    private var mSourcesByNewsCategory: LiveData<List<Sources>>? = null

    private val mTopHeadlines: List<Articles>? = null
    private val mTopHeadlinesByCountryCategory: List<Articles>? = null
    private val mTopHeadlinesByCountry: List<Articles>? = null
    private val mTopHeadlinesBySearch: List<Articles>? = null

    val dataByFavorite: LiveData<List<Articles>>
        get() {
            mNewsByFavorite = articleAccessObject!!.fetchDataByFavorite()
            return mNewsByFavorite!!
        }
    val dataBySetToRead: LiveData<List<Articles>>
        get() {
            mNewsBySetToRead = articleAccessObject!!.fetchDataByToRead()
            return mNewsBySetToRead!!
        }


    init {
        initializeDatabaseValues(application)

    }

    fun initializeDatabaseValues(application: Application) {
        val appDatabase = AppDatabase.getDatabase(application)
        articleAccessObject = appDatabase?.articleAccessObject()
        sourcesAccessObject = appDatabase?.sourcesAccessObject()
        sources = sourcesAccessObject!!.fetchAllData()
        countries = sourcesAccessObject!!.fetchCountryLists()
        allArticlesNoQuery = articleAccessObject!!.fetchAllData()
        newsCategories = sourcesAccessObject!!.fetchCategoryList()

    }

    fun initializeDatabaseValues(application: Application, query: String) {

        val database = AppDatabase.getDatabase(application)
        articleAccessObject = database?.articleAccessObject()
        sourcesAccessObject = database?.sourcesAccessObject()
    }

    fun getmAllArticles(): LiveData<List<Articles>>? {

        return mAllArticles
    }

    fun getSourcesByNewsCategory(categoryQuery: String): LiveData<List<Sources>> {

        mSourcesByNewsCategory = sourcesAccessObject!!.fetchNewsNamesByCategory(categoryQuery)
        return mSourcesByNewsCategory!!
    }

    fun getArticlesByTitle(context: Context, title: String): LiveData<List<Articles>> {

        return mAllArticlesByTitle!!

    }

    fun getArticlesByDomain(context: Context, domain: String): LiveData<List<Articles>> {
        val domain = domain
        mAllArticlesByDomain = articleAccessObject!!.fetchDataByDomain("%$domain%")

        return mAllArticlesByDomain!!

    }

    fun insert(articles: List<Articles>) {
        insertAsyncTask(articleAccessObject!!).execute(articles)

    }

    fun insertFavorite(setToFavorite: Boolean, title: String) {

        val hashMap = HashMap<String, Any>()
        hashMap.put("favorite", setToFavorite)
        hashMap.put("title", title)
    }

    fun insertSetToRead(setToReadValue: Boolean, title: String) {
        val hashMap = HashMap<String, Any>()
        hashMap.put("toReadValue", setToReadValue)
        hashMap.put("title", title)
    }

    fun getArticlesByCountry(context: Context, country: String): LiveData<List<Articles>> {

        return mArticlesByCountry!!
    }

    class insertAsyncTask internal constructor(private val mAsyckTaskAccessObject: ArticleAccessObject) : AsyncTask<List<Articles>, Void, Void>() {

        override fun doInBackground(vararg params: List<Articles>): Void? {

            for (index in 0 until params[0].size) {
                mAsyckTaskAccessObject.createDataIfNotExists(params[0][index])
            }

            return null
        }

    }

    class insertDataFromCountry
    @Inject internal constructor(private val context: Context,
                                 private val queryValue: String,
                                 networkService: NetworkService,
                                 private val mInsertByCountryAccessObject: ArticleAccessObject)
        : AsyncTask<String, Void, Void>() {
        private var articlesByCountry: List<Articles>? = null
        private val modifiedArticleList: List<Articles>? = null

        override fun doInBackground(vararg strings: String): Void? {
            if (strings != null) {

                val connectivityManager = context
                    .getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager

                val activeNetwork = connectivityManager.activeNetworkInfo

                val isConnected = activeNetwork != null && activeNetwork.isConnectedOrConnecting

                return null

            }


            class insertDataFromTitle
            @Inject
            internal constructor(private val context: Context,
                                 private val queryValue: String,
                                 networkService: NetworkService,
                                 private val mInsertByTitleAccessObject: ArticleAccessObject) : AsyncTask<String, Void, Void>() {
                private var articlesByTitle: List<Articles>? = null

                override fun doInBackground(vararg strings: String): Void? {

                    return null
                }

            }

            class insertDataFromDomain
            @Inject
            internal constructor(private val context: Context,
                                 domain: String,
                                 networkService: NetworkService,
                                 private val mInsertByTitleAccessObject: ArticleAccessObject) : AsyncTask<String, Void, Void>() {

                private var queryValue: String? = null
                private var articlesByDomain: List<Articles>? = null

                init {
                    this.queryValue = domain

                    val pattern = Pattern.compile("(https?://)([^:^/]*)(:\\d*)?(.*)?")
                    val matcher = pattern.matcher(domain)
                    matcher.find()
                    this.queryValue = matcher.group(2).substring(4)
                }

                override fun doInBackground(vararg strings: String): Void? {
                    if (strings != null) {

                        val connectivityManager = context
                            .getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager

                        val activeNetwork = connectivityManager.activeNetworkInfo

                        val isConnected = activeNetwork != null && activeNetwork.isConnectedOrConnecting

                    }
                    return null
                }
            }
            return null
        }
    }
}

